src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,117 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (ShouldRestoreNullValues())
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties;
+
+                        if (resourceState.SelectedProperties == SelectedPropertiesNode.EntireSubtree)
+                        {
+                            selectedProperties = GetSelfAndBaseTypeProperties(edmStructuredType);
+                        }
+                        else

```

Comment:
This case is handled already in GetSelectedProperties()

Reply:
Not sure what you meant. It is handled here by GetSelectedProperties().

---
In reply to: [198062679](https://github.com/OData/odata.net/pull/1106#discussion_r198062679) [](ancestors = 198062679)

Reply:
In the above conditional you explicitly check for resourceState.SelectedProperties==SelectedPropertiesNode.EntireSubtree.  If not SelectedPropertiesNode.EntireSubtree, you call GetSelectedProperties, which also checks for SelectionType.EntireSubtree (along with this.hasWildcard) and treats (almost) the same way (by returning all of the properties of the structured type).  You should be able to remove the check here for SelectedPropertiesNode.EntireSubtree, and just invoke GetSelectedProperties.

---
In reply to: [209784487](https://github.com/OData/odata.net/pull/1106#discussion_r209784487) [](ancestors = 209784487,198062679)

Reply:
I see, though that your comment was related to the else branch at L2291. Removed the if-branch above.

---
In reply to: [213508907](https://github.com/OData/odata.net/pull/1106#discussion_r213508907) [](ancestors = 213508907,209784487,198062679)